gslab_scons/provenance.py

```diff
@@ -0,0 +1,195 @@
+            size = stat.st_size
+            mtime = datetime.datetime.fromtimestamp(stat.st_mtime).strftime('%Y-%m-%d %H:%M:%S')
+            # Incorporate file info into directory info.
+            total_size += size            
+            num_files += 1
+            last_mtime = max(last_mtime, mtime)
+            # Optional detailed file information
+            if include_details and (not detail_limit or num_files <= detail_limit): 
+                line = '%s|%s|%s' % (path, size, mtime)
+                if include_checksum:
+                    with open(path, 'rU') as f:
+                        checksum = str(mmh3.hash128(f.read(), 2017))
+                    line = '%s|%s' % (line, checksum)
+                file_details.append(line)
+   
+    return total_size, num_files, last_mtime, file_details, dirs
+
+
+def write_heading(start_path, provenance_path):
+    '''
+    Write standard heading for provenance: what and for where it is. 
+    '''
+    out = '*** GSLab directory provenance ***\ndirectory: %s\n' % os.path.abspath(start_path) 
+    with open(provenance_path, 'wb') as f:
+        f.write(out)
+
+    return None
+
+
+def write_directory_info(provenance_path, total_size, num_files, last_mtime):
+    '''
+    Write directory-level information to provenance. 
+    '''
+    out = 'total bytes: %s\n' % total_size + \
+          'number of files: %s\n' % num_files + \
+          'most recent modification time: %s\n' % last_mtime
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** Directory information\n')
+        f.write(out)
+
+    return None
+
+
+def write_readme(readme_path, provenance_path):
+    '''
+    Writes readme to provenance.
+    '''
+    with open(readme_path, 'rU') as f:
+        out = '%s' % f.read()
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** README verbatim\npath: %s\n' % os.path.abspath(readme_path))
+        f.write(out)
+
+    return None
+
+
+def write_detailed_info(provenance_path, details):
+    '''
+    Writes file-level information to provenance.
+    '''
+    out = '%s\n' % '\n'.join(details)
+    with open(provenance_path, 'ab') as f:
+        f.write('\n*** File information\n')
+        f.write(out)
+
+    return None
+
+
+def write_ending(provenance_path):
+    with open(provenance_path, 'ab') as f:
+        f.write('*' * 80 + '\n\n')
+    
+    return None
+
+
+def append_sub_provenance(root_provenance = './provenance.log',
+                          external_provenance = [],
+                          excluded_dirs = [],
+                          find_for_me = False,
+                          start_path = '.',
+                          verbose = False):
+    
+    files = misc.make_list_if_string(external_provenance)
+    if files == [] or find_for_me == True:
+        pattern = 'provenance*.log'
+        files += misc.finder(start_path, pattern, excluded_dirs)
+    files = sorted(files)
+    
+    
+    with open(root_provenance, 'a') as root_f:
```

Comment:
Open in `'ab'`.

Reply:
Move within the for loop over files (just above `if warn == True` and at same indentation level as opening of each sub-provenance). Collect all lines that write to the provenance file in one place.

Reply:
No idea what you mean here

Reply:
Can I just come over and show you?